### PR TITLE
gh-1090 Rewrite of the passcode auth

### PR DESCRIPTION
**Pull request by DmitryBespalov**
Rewrite of the passcode logic as it was not handling the error states properly.

### CODE REVIEW

File: Multisig/App/SceneDelegate.swift [hunk 1]
```diff
@@ -129,16 +129,22 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
...
     private func showMainWindow() {
+        if snackbarViewController.view.window != mainWindow, let window = mainWindow {
+            window.addSubviewAlwaysOnTop(snackbarViewController.view)
+        }
         mainWindow?.makeKeyAndVisible()
         windowState = .main
     }
 
     private func showPrivacyWindow() {
+        if snackbarViewController.view.window != privacyProtectionWindow, let window = privacyProtectionWindow {
```
Comment (DmitryBespalov) [inline Multisig/App/SceneDelegate.swift:140]: allows to show error messages when entering passcode on app login

File: Multisig/Logic/AuthenticationController.swift [hunk 2]
```diff
@@ -142,55 +141,158 @@ class AuthenticationController {
...
+            App.shared.snackbar.show(error: gsError)
+            completion(.failure(gsError))
+
+        case .success:
+            context.evaluate(policy: policy, reason: reason) { result in
+
+                DispatchQueue.main.async {
+                    switch result {
+                    case .success:
+
```
Comment (sche) [inline Multisig/Logic/AuthenticationController.swift]: extra space

File: Multisig/Logic/AuthenticationController.swift [hunk 3]
```diff
@@ -142,55 +141,158 @@ class AuthenticationController {
...
-                completion(result)
+    func activateBiometrics(completion: @escaping (Result<Void, Error>) -> Void) {
+        evaluate(policy: .deviceOwnerAuthenticationWithBiometrics,
+                 reason: "Enable login with biometrics",
+                 showsFallback: false,
+                 errorConverter: GSError.BiometryActivationError.init(reason:)) { result in
+
+            switch result {
+            case .success:
+
```
Comment (sche) [inline Multisig/Logic/AuthenticationController.swift]: extra space

File: Multisig/Logic/AuthenticationController.swift [hunk 4]
```diff
@@ -142,55 +141,158 @@ class AuthenticationController {
         }
     }
 
-    var isBiometryPossible: Bool {
-        guard let user = user else { return false }
-        return (try? accessService.isAuthenticationMethodPossible(userID: user.id, method: .biometry)) ?? false
```
Comment (sche) [inline Multisig/Logic/AuthenticationController.swift:147]: Do you plan to extend the library later?

**Review by MouazAlzahabi — APPROVED** [review]

**Review by sche — APPROVED** [review]

File: Multisig/Logic/AuthenticationController.swift (hunk 4 above)
Comment (DmitryBespalov) [reply Multisig/Logic/AuthenticationController.swift:147]: no, I even plan to stop using it as it is not needed anymore

**Pull request merged**